### PR TITLE
Allow to apply opacity to all background colors

### LINKED ISSUE #741: Transparency improvements
Based on discussion in #331, we would like to do one or both of the following:

- Add an option to make non-default background colors opaque
- Add an option to fonts to customize their opacity.

### PR DESCRIPTION
**Pull request by satlank**
Alacritty currently only applies the value of `background_opacity` to
the default background and not to all (e.g. #807, addressed in #847).
However this is not necessarily always desired (e.g. #1082, #1180).

This patch adds a configuration option and logic (by partly re-instating
code removed in #847) to toggle the behavior between either applying
only to the default background color, or all background colors.  For
example iterm2 also has this option and the naming of the toggle is
borrowed from there.

Fixes #741.

### CODE REVIEW

**Comment by kchibisov** [comment]
I'm not entirely oposed to that feature knowing that certain terminals indeed apply opacity to everything. However I don't like the implementation, you don't need to change anything in rendering part at all. The thing you want to change is to return 0 in https://github.com/alacritty/alacritty/blob/5ee7ae8a274fcc58754aa5a91a14b1dedbb87583/alacritty_terminal/src/term/mod.rs#L364 when the config option is activated.

**Comment by satlank** [comment]
Thanks for looking at the PR and the suggestion (which sounds much simpler). I may be missing something, but it appears to be more complicated than just returning 0, as that seems to make all backgrounds to disappear. E.g. in the attached image the left terminal is with the patch in this PR, the right one from master with the shown diff.

image

Any suggestion where I should be looking?

**Comment by kchibisov** [comment]
Ah, yeah, you need to return `background_opacity()` from `compute_bg_alpha()` and do in text.f.glsl `color = vec4(bg.rgb, 1.0) * bg.a` in a place where you have added multiply by the value from uniform in your patch.

**Comment by satlank** [comment]
Right, okay, that seems to work, albeit it needs to be `color = vec4(bg.rgb, 1.0) * bg.a / 255.0` in `text.f.glsl`, it appears `bg.a` gets scaled to [0,255] somewhere in between returning a float [0,1.0] from `compute_bg_alpha()` and using it in the GSLS program. I am not sure about performance impacts at that point, naively a division instead of a simple multiply seems undesirable and maybe injecting the relevant multiplier directly as done originally is simpler?

It is going to be a bit messy with the config settings: I had added the `keep_background_colors_opaque` to `struct UIConfig` to keep it with `background_opacity`, but that isn't (intentionally, I suppose) accessible from `alacritty_terminal`. However if I move it to `struct Config<T>` in `alacritty_terminal/src/config/mod.rs` the background opacity isn't accessible for use by `RenderApi.clear()` in `alacritty/src/renderer/mod.rs`.

Any thoughts on how this should be structured?

**Comment by kchibisov** [comment]
>gets scaled to [0,255] somewhere in between returning a float [0,1.0] from

it get scaled when pushing instance data in render iirc.

> naively a division instead of a simple multiply seems undesirable and maybe injecting the relevant multiplier directly as done originally is simpler?

well, you need to adjust vertex shader to make division, forgot about that part. it's already there, just need to include alpha in it as well.

 
>It is going to be a bit messy with the config settings: I had added the keep_background_colors_opaque to struct UIConfig to keep it with background_opacity, but that isn't (intentionally, I suppose) accessible from alacritty_terminal. However if I move it to struct Config<T> in alacritty_terminal/src/config/mod.rs the background opacity isn't accessible for use by RenderApi.clear() in alacritty/src/renderer/mod.rs.

that's the most annoying part here tbqh. In general, I'd like to see splitting out the Renderable iterator from alacritty_terminal, since rgb conversions must be done by alacritty, since it makes it really hard to use the lib in cases like embedding in other applications, etc,  however it's not that simple to approach, but I can't think of some simple solution to that problem of passing opacity around.

@chrisduerr do you have any opinion on 'extracting iterator' from `alacritty_terminal` or changing things that it returns an iterator over visible chars(with touching them) and you in alacritty builds another iterator out of that or maps its elements?

The original solution with uniform isn't that bad, but I just feel like it's too much hasstle for opacity option, knowing that we already have one and making all things transparent is subjecive.

**Comment by chrisduerr** [comment]
> @chrisduerr do you have any opinion on 'extracting iterator' from alacritty_terminal or changing things that it returns an iterator over visible chars(with touching them) and you in alacritty builds another iterator out of that or maps its elements?

I feel like any suggestion in this direction would be out of scope for this issue? If we want to do that, it should be done separately. But in general nothing stops you from having unmapped colors returned by the cell iterator, right?

**Comment by kchibisov** [comment]
>I feel like any suggestion in this direction would be out of scope for this issue?

Yeah, it'd a bit out of scope for that issue.

>But in general nothing stops you from having unmapped colors returned by the cell iterator, right?

Right.

**Comment by satlank** [comment]
Ok, I am not sure I am following the intricacies of where cell iterators ought to be (and the finer points of what belongs into `alacritty` and what into `alacritty_terminal`). I does sound like the opacity handling should all happen in `alacritty`, but it would require refactoring to move the relevant bits from `alacritty_terminal` to `alacritty` first?

Looking at the configuration conundrum again, there is a way to leave `background_opacity` in `UIConfig`: Instead of introducing a toggle (`keep_background_colors_opaque`) I can add another opacity value `non_default_background_opacity` (needs a better name, naming is hard, bear with me) which would be attached to `Config<T>`, defaulted to 1.0 and used instead of the hard-coded 1.0 in `compute_bg_alpha()` - it should be set to the same value as `background_opacity` or 1.0 to recover the desired behaviour. But that doesn't look particularly appealing either.

Maybe it's best to not merge this feature then and revisit it when the separation is cleaner?

**Comment by chrisduerr** [comment]
> It does sound like the opacity handling should all happen in alacritty, but it would require refactoring to move the relevant bits from alacritty_terminal to alacritty first?

Yes. Currently it is not done ***entirely*** in `alacritty`, which it should.

> Looking at the configuration conundrum again, there is a way to leave background_opacity in UIConfig

Just to be clear, when it comes to configuration options there is no other choice. It must be in the UIConfig, we can't put UI config stuff in the `alacritty_terminal` config. Since moving it after the fact is just going to be a massive pain. In fact the `background_opacity` option should be on `window`.

> Maybe it's best to not merge this feature then and revisit it when the separation is cleaner?

I'm open to PRs that separate the code. It's likely the easiest solution. I'm just saying that it shouldn't be done in this PR.

**Comment by kchibisov** [comment]
Well, as an alternative solution we can do something like
```diff
diff --git a/alacritty/res/text.f.glsl b/alacritty/res/text.f.glsl
index c0c1bd1b..de0e9fa9 100644
--- a/alacritty/res/text.f.glsl
+++ b/alacritty/res/text.f.glsl
@@ -16,7 +16,7 @@ void main()
             discard;

         alphaMask = vec4(1.0);
-        color = vec4(bg.rgb, 1.0);
+        color = vec4(bg.rgb, 1.0) * bg.a;
     } else {
         if (fg.a != 0.0) {
             // Color glyphs, like emojis.
diff --git a/alacritty/res/text.v.glsl b/alacritty/res/text.v.glsl
index 0f5fd3f9..bff3f42d 100644
--- a/alacritty/res/text.v.glsl
+++ b/alacritty/res/text.v.glsl
@@ -56,6 +56,6 @@ void main()
         TexCoords = uvOffset + position * uvSize;
     }

-    bg = vec4(backgroundColor.rgb / 255.0, backgroundColor.a);
+    bg = backgroundColor / 255.;
     fg = vec4(textColor.rgb / 255.0, textColor.a);
 }
diff --git a/alacritty/src/renderer/mod.rs b/alacritty/src/renderer/mod.rs
index 6b27a837..3d66b96c 100644
--- a/alacritty/src/renderer/mod.rs
+++ b/alacritty/src/renderer/mod.rs
@@ -491,7 +491,7 @@ impl Batch {
             bg_r: cell.bg.r,
             bg_g: cell.bg.g,
             bg_b: cell.bg.b,
-            bg_a: (cell.bg_alpha * 255.0) as u8,
+            bg_a: cell.bg_alpha as u8,
             multicolor: glyph.multicolor,
         });
     }
@@ -995,7 +995,13 @@ impl<'a> RenderApi<'a> {
     }

     #[inline]
-    fn add_render_item(&mut self, cell: RenderableCell, glyph: &Glyph) {
+    fn add_render_item(&mut self, mut cell: RenderableCell, glyph: &Glyph) {
+        if self.config.keep_background_colors_opaque {
+            cell.bg_alpha *= 255.;
+        } else {
+            cell.bg_alpha *= 255. * self.config.background_opacity();
+        }
+
         // Flush batch if tex changing.
         if !self.batch.is_empty() && self.batch.tex != glyph.tex_id {
             self.render_batch();

```

**Comment by satlank** [comment]
Thanks for the clarifications! I'll try to update the PR tonight, I am on the wrong machine for most of the day, your alternative solution looks like a good middle ground.

**Comment by satlank** [comment]
Ok, all updated. It certainly is a neater diff now, thanks again for the improvement suggestions.

File: alacritty.yml
Comment (chrisduerr) [inline alacritty.yml:328]: I think the naming of this option isn't ideal, since it says little about what it does in its name and I'd advocate for moving this option to `window` instead of keeping it at the top level. Together with `background_opacity` probably.

File: alacritty/src/renderer/mod.rs
Comment (chrisduerr) [inline alacritty/src/renderer/mod.rs:1003]:
```suggestion
        cell.bg_alpha *= 255.;
        if !self.config.keep_background_colors_opaque() {
            cell.bg_alpha *= self.config.background_opacity();
        }
```

**Review by chrisduerr — CHANGES_REQUESTED** [review]

File: alacritty/src/renderer/mod.rs
Comment (satlank) [reply alacritty/src/renderer/mod.rs:1003]: Oh, yes, will do that.

File: alacritty.yml
Comment (satlank) [reply alacritty.yml:328]: I am open to suggestions for a better name. My 'inspiration' for naming it that way came from iterm (not that that is a strong argument):
image

Happy to move it to the window section, do you want me to move only the new option for now so that it is in the right place to start with or also `background_opacity`? If the latter what is the procedure to deprecate configuration options or is a breaking change fine?

File: alacritty/src/renderer/mod.rs
Comment (satlank) [reply alacritty/src/renderer/mod.rs]: That's done.

File: alacritty.yml
Comment (chrisduerr) [reply alacritty.yml:328]: > My 'inspiration' for naming it that way came from iterm (not that that is a strong argument):

I mean that's definitely a solid argument, thanks for letting me know. I'm struggling to suggest a much better alternative though. My main problem is that the `keep_` prefix doesn't really fit in well with our configuration options. Maybe something more like `opaque_colored_backgrounds`?

If you have some other suggestions, please do let me know.

> Happy to move it to the window section, do you want me to move only the new option for now so that it is in the right place to start with or also background_opacity? If the latter what is the procedure to deprecate configuration options or is a breaking change fine?

I think I'd move both, just to have it all in place. Though I'm not entirely sure if this should be under `colors` or `window`. @kchibisov any thoughts?

Comment (kchibisov) [reply alacritty.yml:328]: it should be under window, I'd say.

Comment (kchibisov) [reply alacritty.yml:328]: >I mean that's definitely a solid argument, thanks for letting me know. I'm struggling to suggest a much better alternative though. My main problem is that the keep_ prefix doesn't really fit in well with our configuration options. Maybe something more like opaque_colored_backgrounds?

We can have something like `apply_opacity_to_background_colors: false` setting, but I don't really mind original option.

**Review by kchibisov — CHANGES_REQUESTED** [review]
Could you move your new config option to `window` section, and also `background_opacity` (with adding deprecation for that one)?

File: alacritty.yml
Comment (chrisduerr) [reply alacritty.yml:328]: That's a tad long.

**Comment by chrisduerr** [comment]
Okay, I've taken a bit of a look at this myself and I think the following is what I would recommend.

<details>

```diff
diff --git a/alacritty.yml b/alacritty.yml
index ac4477f..8a02ca6 100644
--- a/alacritty.yml
+++ b/alacritty.yml
@@ -21,6 +21,12 @@
   #TERM: alacritty
 
 #window:
+  # Background opacity
+  #
+  # Window opacity as a floating point number from `0.0` to `1.0`.
+  # The value `0.0` is completely transparent and `1.0` is opaque.
+  #background_opacity: 1.0
+
   # Window dimensions (changes require restart)
   #
   # Specified in number of columns/lines, not pixels.
@@ -172,6 +178,13 @@
 
 # Colors (Tomorrow Night)
 #colors:
+  # Opaque backgrounds with transparency enabled
+  #
+  # Whether or not `background_opacity` applies to all backgrounds or only the
+  # default background. When set to `false` all cells will be transparent
+  # regardless of their background color.
+  #opaque_background_colors: true
+
   # Default colors
   #primary:
   #  background: '#1d1f21'
@@ -316,12 +329,6 @@
   #
   #command: None
 
-# Background opacity
-#
-# Window opacity as a floating point number from `0.0` to `1.0`.
-# The value `0.0` is completely transparent and `1.0` is opaque.
-#background_opacity: 1.0
-
 #selection:
   # This string contains all characters that are used as separators for "semantic words" in Alacritty.
   #semantic_escape_chars: ",│`|:\"' ()[]{}<>\t"
```

</details>

This should be a nice descriptive name and simple to understand documentation. Having one on color while the other on window I'd explain since things like blur/transparency are usually understood on a window level, but the transparency of individual colors is not really related to the window.

Let me know what you think.

**Comment by satlank** [comment]
> This should be a nice descriptive name and simple to understand documentation. Having one on color while the other on window I'd explain since things like blur/transparency are usually understood on a window level, but the transparency of individual colors is not really related to the window.
> 
> Let me know what you think.

I like the new naming, so did that. I have left both options together in the config though for two reasons, a) it makes more sense to me to have that options with the opacity settings as it toggles how opacity is applied, and b) when moved to colors it isn't accessible in the render (that only has access to the UIConfig). Can change the logic on where it is applied to address b) I suppose.

File: alacritty.yml
Comment (satlank) [reply alacritty.yml]: I've changed it your suggestion.

**Comment by satlank** [comment]
> Could you move your new config option to window section, and also background_opacity (with adding deprecation for that one)?

That's done.

File: alacritty/src/config/window.rs
Comment (satlank) [inline alacritty/src/config/window.rs:16]: Needed to drop `Eq` as `Percentage` is effectively a `f32` and thus doesn't have `Eq`.

File: alacritty/src/config/ui_config.rs
Comment (satlank) [inline alacritty/src/config/ui_config.rs:49]: `pub` because otherwise the deprecation check can't access it.

**Comment by chrisduerr** [comment]
> it makes more sense to me to have that options with the opacity settings as it toggles how opacity is applied

It has nothing to do with the window though.

> when moved to colors it isn't accessible in the render (that only has access to the UIConfig). Can change the logic on where it is applied to address b) I suppose.

That's something that can be easily changed.

**Comment by kchibisov** [comment]
>That's something that can be easily changed.

Oh, but colors is a part of a alacritty_terminal config, so and moving opacity there is a bit of a problem tbqh, since it has nothing to do with a terminal. In general colors should be in UIConfig, but it's not that simple to do.

What do you think about having them at a top level for a while? And once we move colors to UIConfig we can move opacity things as well?

Alternatively we can go to a thing we've used in a past with `Config<UIConfig>` in a renderer, so not really a big deal, and it's something simple to change.

**Comment by chrisduerr** [comment]
> Oh, but colors is a part of a alacritty_terminal config

I mean that'll change in the future anyways.

> What do you think about having them at a top level for a while?

I think that introducing a config option knowing fully well it won't stay where it is will only introduce unnecessary churn for users.

> Alternatively we can go to a thing we've used in a past with Config<UIConfig> in a renderer, so not really a big deal, and it's something simple to change.

There's absolutely nothing wrong with passing the full config to the renderer, it changes nothing. The only thing we cannot do is pass the UIConfig to the terminal.

**Comment by satlank** [comment]
Sorry, have been snowed under the last couple of days. I'll try to get back to this tomorrow.

**Comment by caseyavila** [comment]
Any updates/progress for this feature?

**Comment by satlank** [comment]
> Any updates/progress for this feature?

Yeah, sorry, half-forgot about this. I'll try to get back to this over the holidays, need to update my local binaries with master anyway.

**Comment by ric2work** [comment]
Hi

**Comment by satlank** [comment]
Ok, I think this has addressed the changes requested (and I've also rebased onto master dealing with the merge conflicts).

@chrisduerr, apologies for the delay in getting back to this. Is there anything else that needs changing?

File: alacritty.yml
Comment (chrisduerr) [inline alacritty.yml:297]:
```suggestion
  # Whether or not `background_opacity` applies to all backgrounds or only to the
  # default background. When set to `false` all cells will be transparent
  # regardless of their background color.
```

File: alacritty.yml
Comment (chrisduerr) [inline alacritty.yml:298]: It has been a while and I don't completely remember our previous discussions, but `opaque_colored_backgrounds` might be an alternative name for this that is a little more clear? I'm not sure though, I think both would probably be fine.

File: alacritty/res/text.f.glsl
Comment (chrisduerr) [inline alacritty/res/text.f.glsl:21]: I'm not sure this logic is correct? Just looking at the results, it seems like `color = vec4(bg.rgb, bg.a);` gives me results closer to what I would expect?

Neither of them seem to actually be correct though, since it seems like they aren't independent from the actual background color? I'd expect a transparent cell background with the same hex color as the background to have the exact same color, but it doesn't. This is likely because the cell's background is getting merged with the background of the terminal.

File: alacritty/src/config/ui_config.rs
Comment (chrisduerr) [inline alacritty/src/config/ui_config.rs:44]: This annotation isn't necessary anymore, the new `#[config(deprecated)]` annotation is sufficient to fiend deprecated fields.

File: alacritty/src/config/ui_config.rs
Comment (chrisduerr) [inline alacritty/src/config/ui_config.rs:72]: This might work:

```rust
self.background_opacity.map(Percentage::as_f32).unwrap_or_else(|| self.window.background_opacity())
```

File: alacritty/src/renderer/mod.rs
Comment (chrisduerr) [inline alacritty/src/renderer/mod.rs:812]: This can probably stay where it was? Doesn't matter by which value we multiply first, so we can keep it in the 1.0 to 0.0 range for now.

File: alacritty/src/renderer/mod.rs
Comment (chrisduerr) [inline alacritty/src/renderer/mod.rs:815]: Is this the right place for this? If `cell` already has a `bg_alpha` field, shouldn't this property be applied when this field is initialized? Doing it right here seems somewhat arbitrary?

File: alacritty_terminal/src/config/colors.rs
Comment (chrisduerr) [inline alacritty_terminal/src/config/colors.rs:34]:
```suggestion
            opaque_background_colors: true,
            primary: Default::default(),
            cursor: Default::default(),
            vi_mode_cursor: Default::default(),
            selection: Default::default(),
            normal: Default::default(),
            bright: Default::default(),
            dim: Default::default(),
            indexed_colors: Default::default(),
            search: Default::default(),
```

Putting non-default fields at the top makes it easy to see why there is a custom `Default` implementation.

**Review by chrisduerr — CHANGES_REQUESTED** [review]
Thanks for picking this back up.

File: alacritty.yml
Comment (satlank) [reply alacritty.yml:298]: I don't have strong feelings either way. Whatever you prefer.

File: alacritty/src/config/ui_config.rs
Comment (satlank) [reply alacritty/src/config/ui_config.rs:44]: Ah, neat, also removes my typo o_O..

File: alacritty/src/config/ui_config.rs
Comment (satlank) [reply alacritty/src/config/ui_config.rs:72]: Yes, looks like it would after some mental gymnastics. Not too used to the unwrap-chain bits..  Is that any clearer though?

File: alacritty.yml
Comment (chrisduerr) [reply alacritty.yml:298]: Let's just leave it as-is for now. I don't really think it's a significant improvement.

File: alacritty/src/renderer/mod.rs
Comment (satlank) [reply alacritty/src/renderer/mod.rs:812]: I think there was a reason to do this, but I don't recall. I'll try whether I can figure out again why I changed that here.

File: alacritty/src/config/ui_config.rs
Comment (chrisduerr) [reply alacritty/src/config/ui_config.rs:72]: Yeah, I'd recommend this over the existing code.

File: alacritty/res/text.f.glsl
Comment (satlank) [reply alacritty/res/text.f.glsl:21]: Hm. I vaguely remember thinking about this and preferring it to darken the colours as well when applying the opacity. I'll go through this again and see if I can retrace my steps.

File: alacritty/src/renderer/mod.rs
Comment (satlank) [reply alacritty/src/renderer/mod.rs:815]: Well, I think I basically followed https://github.com/alacritty/alacritty/pull/4196#issuecomment-686057677 with where to put the scaling. Maybe there is a better way to put it, since I am anyway going to retrace my steps to address the other options I'll bear this is in mind an see if there is a more logical place.

Comment (chrisduerr) [reply alacritty/src/renderer/mod.rs:815]: Ah, I've previously said the following:



> > It does sound like the opacity handling should all happen in alacritty, but it would require refactoring to move the relevant bits from alacritty_terminal to alacritty first?
>
> Yes. Currently it is not done entirely in alacritty, which it should.

This is technically correct, but fact is that opacity handling happens in `alacritty_terminal` already. So if we want to move it outside of `alacritty_terminal`, it should probably all be moved. I believe that's probably what I was thinking of back then too. Generally though that's a lot of work, so I'd be fine with putting it in `alacritty_terminal` for now, you should be able to find the other alpha stuff in `RenderableCell::new` I think.

File: alacritty/res/text.f.glsl
Comment (chrisduerr) [reply alacritty/res/text.f.glsl:21]: As I've stated, I think both are incorrect. Generally the ideal solution would be to do this *correctly*, but I haven't looked into that so I can't show you a solution for that.

But at least `vec4(bg.rgb, bg.a)` looks like it is a little more correct at least on a dark background.

Comment (satlank) [reply alacritty/res/text.f.glsl:21]: Okay, I think the reason I went with the multiplier on all instead of just setting alpha was that the latter messed up the background more visibly when entering tmux (it's a bit difficult to tell, but the issue is the more whiteish background in tmux compared to the non-tmux in the last two screenshots below - the opacity is set to 0.4 to have a more extreme impact, I normally run with 0.9).

I think you are right that there is something fishy going on, it does look like that tmux/no-tmux does look slightly different in my implementation as well. But other than disecting the complete rendering path I don't know how to do it properly - so certainly won't come up with a solution today and possibly not this year either. I'll poke around as I get time over the next days. I *think* that my original implementation (which just backed out changes that switch the original opacity-to-all to the now-in-master opacity-only-to-default-bg) didn't have this problem, so maybe that's a place to start. 

master:
master
master tmux:
master tmux
this branch:
branch
this branch tmux:
branch-tmux
this branch only alpha:
only-alpha
this branch only alpha tmux:
only-alpha-tmux

File: alacritty/src/config/ui_config.rs
Comment (satlank) [reply alacritty/src/config/ui_config.rs]: Okay, that's done.

File: alacritty/src/config/ui_config.rs
Comment (satlank) [reply alacritty/src/config/ui_config.rs]: Dropped.

File: alacritty/res/text.f.glsl
Comment (satlank) [reply alacritty/res/text.f.glsl:21]: Ah, the difference in bg between tmux/non-tmux case is because of how the background opacity is currently done in master (https://github.com/alacritty/alacritty/blob/master/alacritty/src/renderer/mod.rs#L761-L766), it doesn't just set the alpha channel, it also scales the RGB channels. I don't have an opinion on which is preferable, but changing both `clear` to only set the alpha channel and `color = vec4(bg.rgb, bg.a)` I don't see a difference in bg colour between my tmux/non-tmux cases.
So if you are happy to change how the opacity is applied also in the 'only for the default bg' case, then it works to just apply it to the alpha channel.
For how I am using the opacity setting it doesn't make much of a difference. But for a smaller opacity setting the difference is much more notable, though I find it hard to tell which version 'looks better'.

Comment (chrisduerr) [reply alacritty/res/text.f.glsl:21]: I have no problem with changing things as long as it works properly. There shouldn't be much "opinion" involved. 50% opacity should be a pretty clear instruction usually.

Comment (satlank) [reply alacritty/res/text.f.glsl]: ¯\_(ツ)_/¯
Changed to only apply the opacity to the alpha channel.

File: alacritty/src/renderer/mod.rs
Comment (satlank) [reply alacritty/src/renderer/mod.rs:857]: I think I see where to do that in alacritty_terminal. However while I can get to `config.colors.opaque_background_color` I am not sure that `config.ui_config.background_opacity()` is accessible from there - at least the compiler keeps shouting at me? I suppose I could move the `background_opacity` to `color`, but given that you were keen on moving it to `window` I am not sure that is desirable (plus I am not sure how to support the previous location of that field)?

Comment (chrisduerr) [reply alacritty/src/renderer/mod.rs:856]: Fair enough, I'd hold on this for now then I suppose.

File: alacritty/src/renderer/mod.rs
Comment (satlank) [reply alacritty/src/renderer/mod.rs]: Not sure why that got changed. Seems to be fine to leave that scaling in the batch so moved it back there.

File: alacritty/res/text.f.glsl
Comment (chrisduerr) [reply alacritty/res/text.f.glsl]: You just need to watch out, because the glyph data is not in RGB format.

**Comment by dankamongmen** [comment]
If you're going to do this, I'd appreciate you making it discoverable from clients. Something might be relying on that background color. Overall I think this is a pretty bad idea. What if my program is indicating an error state by throwing a red background on some glyphs? With this patch, as I understand things, you'd throw away that important indicator for...what, exactly?

Kitty does this with one particular color, and I already have to deal with that in Notcurses. Please see this bug:

 https://github.com/dankamongmen/notcurses/issues/1117

and see how this can have unexpected effects (you break any kind of plot or other non-text being drawn with block characters, for instance -- imagine that i'm drawing with half blocks. One half of the cell is my foreground color, one is my background color. You just mapped half of my "pixels" to the same color).

I didn't know this about iTerm2, but I'm going to go file a feature request against them. In the meantime, I feel such a feature is incomplete without:

* a way to discover whether it is enabled, or at least push the current setting onto a stack, and
* a way to disable/enable this behavior at runtime, ideally expressed as a terminfo extension capability

otherwise you're introducing a feature with ramifications the user might not foresee, with no way to work around it in clients that know they want to. please don't, @satlank .

**Comment by caseyavila** [comment]
Well this is set up to be configured in the config file, so not only is it optional, but I suppose one could read the configuration file to figure out whether the feature is toggled.

**Comment by chrisduerr** [comment]
https://github.com/alacritty/alacritty/pull/5297

**Comment by dankamongmen** [comment]
> Well this is set up to be configured in the config file, so not only is it optional, but I suppose one could read the configuration file to figure out whether the feature is toggled.

no other capabilities are discovered in such a manner, for good reason. what if there is no config file? i'd have to know/track the default. what if they're using a config file specified via `--config-file`? how am i supposed to know their `XDG_CONFIG_HOME`? was this a serious suggestion?

**Pull request closed without merging**
